File: day5.py
#!/usr/bin/env python3

from itertools import count

raw_input = []
with open("inputs/day5.txt") as file:
    raw_input = file.readlines()
input_list = list(map(int, raw_input))

idx = 0
length = len(raw_input)
for cnt in count(1):
    data_at_idx = input_list[idx]
    input_list[idx] += 1
    idx += data_at_idx
    if idx >= length:
        print("Output part1: " + str(cnt))
        break

input_list = list(map(int, raw_input))
idx = 0
for cnt in count(1):
    data_at_idx = input_list[idx]
    input_list[idx] += -1 if data_at_idx > 3 else 1
    idx += data_at_idx - 1
    if idx >= length:
        print("Output part2: " + str(cnt))
        break

File: rust_solutions/src/day5.rs
use std::env;
use std::io::{self, BufReader};
use std::io::prelude::*;
use std::fs::File;
use std::time::Instant;


// Day 5 part 2 solution in rust; overhauled performance to a total of ~90ms

fn main() {
    let filename = "../inputs/day5.txt";
    let file = File::open(filename).expect("file not found");
    let buf_reader = BufReader::new(file);
    let lines = buf_reader.lines();

    let mut arr: [isize; 1200] = [0; 1200];
    let length: isize = arr.len() as isize;

    let mut cnt: usize = 0;
    for line in lines {
        arr[cnt] = line.unwrap().parse::<isize>().unwrap();
        cnt += 1usize;
    }
    // just for some time measurement
    let start = Instant::now();

    let mut idx: isize = 0;
    let mut steps: u32 = 0;

    loop {
        steps += 1u32;
        let data_at_idx: isize = arr[idx as usize] as isize;
        if data_at_idx >= 3 {
            arr[idx as usize] -= 1;
        } else{
            arr[idx as usize] += 1;
        }
        idx += data_at_idx;
        if idx >= length {
            println!("Part 2 solution = {}", steps as isize);
            break;
        }
    }

    // printing elapsed time
    let elapsed = start.elapsed();
    println!("Elapsed: {} ms",
            (elapsed.as_secs() * 1_000) + (elapsed.subsec_nanos() / 1_000_000) as u64);
}
